src/Range/Pattern.php

```diff
@@ -143,20 +153,95 @@ public function getAddressType()
         return $this->fromAddress->getAddressType();
     }
 
+    /**
+     * {@inheritdoc}
+     *
+     * @see RangeInterface::getRangeType()
+     */
+    public function getRangeType()
+    {
+        if ($this->rangeType === null) {
+            switch ($this->getAddressType()) {
+                case AddressType::T_IPv4:
+                    // Default is public
+                    $this->rangeType = RangeType::T_PUBLIC;
+                    $reservedRanges = IPv4::getReservedRanges();
+                    break;
+                case AddressType::T_IPv6:
+                    if (Subnet::fromString('2002::/16')->containsRange($this)) {
+                        $this->rangeType = Factory::rangeFromBoundaries($this->fromAddress->toIPv4(), $this->toAddress->toIPv4())->getRangeType();
+
+                        return $this->rangeType;
+                    } else {
+                        // Default is public
+                        $this->rangeType = RangeType::T_RESERVED;
+                        $reservedRanges = IPv6::getReservedRanges();
+                    }
+                    break;
+            }
+
+            // Check if range is contained within an RFC subnet
+            foreach ($reservedRanges as $reservedRange) {
+                if ($reservedRange['range']->containsRange($this)) {
+                    $this->rangeType = $reservedRange['type'];
+                    break;
+                }
+            }
+
+            // Check if public/reserved range contains an RFC subnet
```

Comment:
@Empornium could you explain a bit the logic if this piece of code (from line 191 to line 214)?

Reply:
The default for IPv4 is "Public" and for IPv6 is "Reserved", at this point in the function we've already checked if the supplied pattern is contained in any defined ranges so it should have "Public" or "Reserved" if it is not contained within a range returned by getReservedRanges().

After determining it isn't contained within one of the defined reserved ranges we need to determine if it contains any reserved ranges (with a few exceptions) and return null if it does.

For instance, 192.166.0.0/8 is not contained within any defined ranges and as such will have a RangeType of T_PUBLIC at this point in the function, it does however contain within in 192.168.0.0/16 and so should return null at the end.

The specific exceptions are defined in the various RFCs.

Does that make sense?

Reply:
Thanks for the explanation. I already simplified this approach (see the current code in the master branch), so that it's easier to underatand and to maintain. And I also fixed some errors in your approach (see the changes I did on your tests).

Reply:
Classy.